<a href="https://colab.research.google.com/github/peculab/AI4JUBO/blob/main/JuboDeath_V8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#訓練資料是 mortality_2020_2023_0910/training_data_v2_0917
#外部驗證資料是 mortality_2024_0910/external_validation_0917

#次族群

- <= 85 & > 85
- ADL 變好 & ADL 變差
- 男性 & 女性

In [2]:
!pip install shap plotly xgboost --quiet

In [3]:
!pip uninstall shap -y
!pip install shap --no-deps

Found existing installation: shap 0.48.0
Uninstalling shap-0.48.0:
  Successfully uninstalled shap-0.48.0
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 53.0 MB/s eta 0:00:00


In [4]:
!pip install ace_tools

In [5]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score, accuracy_score
from IPython.display import display
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score, roc_auc_score,
    classification_report, confusion_matrix, mean_absolute_error, r2_score
)

In [6]:
from google.colab import auth
auth.authenticate_user()

import gspread
from google.auth import default
creds, _ = default()

gc = gspread.authorize(creds)

外部資料讀入

In [7]:
# read data and put it in a dataframe
# 在 google 工作表載入外部資料 gsheets

gsheets = gc.open_by_url('https://docs.google.com/spreadsheets/d/1jXuepVDDrnhOKYMqy-RQKeQlWr7atvKdeq_Y0PosMDM/edit?usp=sharing')
worksheet = gsheets.worksheet("external_validation_0917")  # 指定分頁名稱

worksheet = worksheet.get_all_records()
external = pd.DataFrame(worksheet)
external = external.apply(lambda col: pd.to_numeric(col.astype(str).str.replace(',', '').str.strip(), errors='coerce'))
external.head()

H01_NUM  dbname  入家日期  結案日期  死亡標記  觀察天數  性別_is_male  預估年齡  DNR_flag  \
0     1376     NaN   NaN   NaN     0   197           1    77         0   
1     1322     NaN   NaN   NaN     0   327           1    92         0   
2     1319     NaN   NaN   NaN     0   255           1    78         1   
3     1333     NaN   NaN   NaN     0   293           1    82         1   
4     1452     NaN   NaN   NaN     0   341           0    80         1   

   ADL_總分_max  ...  first_ 意識總分  last_ 意識總分  意識總分Max  意識總分_diff  意識分級  使用呼吸輔具  \
0          90  ...         15.0        15.0     15.0        0.0   NaN     0.0   
1          10  ...         12.0        12.0     12.0        0.0   NaN     0.0   
2           5  ...         11.0        10.0     11.0       -1.0   NaN     0.0   
3          10  ...         12.0        12.0     12.0        0.0   NaN     0.0   
4          10  ...         11.0        11.0     11.0        NaN   NaN     0.0   

   first_has_feeding_tube  last_has_feeding_tube  diff_has_feeding_tube  \
0                     0.0                    0.0                    0.0   
1                     0.0                    0.0                    0.0   
2                     0.0                    0.0                    0.0   
3                     0.0                    0.0                    0.0   
4                     0.0                    0.0                    0.0   

   had_fall  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  

[5 rows x 45 columns]

In [8]:
external.columns

Index(['H01_NUM', 'dbname', '入家日期', '結案日期', '死亡標記', '觀察天數', '性別_is_male',
       '預估年齡', 'DNR_flag', 'ADL_總分_max', 'ADL_first_score', 'ADL_last_score',
       'ADL_diff_seq', 'ADL_std', 'ADL_Max', 'ADL_Min', 'ADL_明顯惡化',
       'ADL_first_CouldNot', 'ADL_last_CouldNot', 'disease_first_score',
       'disease_last_score', 'disease_diff_seq', 'disease_std', 'disease_Max',
       'disease_Min', '六個月內住院次數', 'first_has_denture', 'last_has_denture',
       'diff_has_denture', 'first_ 意識_E', 'last_ 意識_E', 'first_ 意識_V',
       'last_ 意識_V', 'first_ 意識_M', 'last_ 意識_M', 'first_ 意識總分', 'last_ 意識總分',
       '意識總分Max', '意識總分_diff', '意識分級', '使用呼吸輔具', 'first_has_feeding_tube',
       'last_has_feeding_tube', 'diff_has_feeding_tube', 'had_fall'],
      dtype='object')

In [9]:
external['呼吸_ADL低'] = ((external['使用呼吸輔具'] == 1) & (external['ADL_總分_max'] < 30)).astype(int)
external['高齡_住院多'] = ((external['預估年齡'] > 85) & (external['六個月內住院次數'] > 2)).astype(int)
external['ADL低_明顯惡化'] = ((external['ADL_總分_max'] < 30) & (external['ADL_明顯惡化'] == 1)).astype(int)
external['DNR_呼吸'] = ((external['DNR_flag'] == 1) & (external['使用呼吸輔具'] == 1)).astype(int)
external['超級重症'] = (
    (external['DNR_flag'] == 1) &
    (external['使用呼吸輔具'] == 1) &
    (external['ADL_總分_max'] < 30) &
    (external['六個月內住院次數'] > 1)
).astype(int)
external["low_adl_no_var"] = ((external["ADL_std"].isnull()) & (external["ADL_總分_max"] < 5)).astype(int)

In [10]:
external.describe(include='all').T

count         mean         std    min     25%  \
H01_NUM                 6339.0  1383.704685  767.419894    4.0  1184.0   
dbname                     0.0          NaN         NaN    NaN     NaN   
入家日期                       0.0          NaN         NaN    NaN     NaN   
結案日期                       0.0          NaN         NaN    NaN     NaN   
死亡標記                    6339.0     0.280959    0.449503    0.0     0.0   
觀察天數                    6339.0   210.107588  109.964750    0.0   127.5   
性別_is_male              6339.0     0.508440    0.499968    0.0     0.0   
預估年齡                    6339.0    78.607825   11.742108    1.0    72.0   
DNR_flag                6339.0     0.456854    0.498174    0.0     0.0   
ADL_總分_max              6339.0    29.695536   31.687105    0.0     0.0   
ADL_first_score         6273.0    26.496094   30.143640    0.0     0.0   
ADL_last_score          6272.0    25.596301   30.464615    0.0     0.0   
ADL_diff_seq            6246.0    -0.898975   16.170515 -100.0     0.0   
ADL_std                 4834.0     5.055448    8.799599    0.0     0.0   
ADL_Max                 6339.0    29.695536   31.687105    0.0     0.0   
ADL_Min                 6339.0    21.894621   28.481462    0.0     0.0   
ADL_明顯惡化                6339.0     0.127149    0.333167    0.0     0.0   
ADL_first_CouldNot      6339.0     0.010412    0.101513    0.0     0.0   
ADL_last_CouldNot       6339.0     0.010569    0.102271    0.0     0.0   
disease_first_score     2503.0     1.690851    0.597390    0.0     1.3   
disease_last_score      2506.0     1.715842    0.558684    0.0     1.3   
disease_diff_seq        2501.0     0.024670    0.297071   -2.1     0.0   
disease_std             1668.0     0.103360    0.191756    0.0     0.0   
disease_Max             2520.0     1.759008    0.584624    0.0     1.4   
disease_Min             2520.0     1.626429    0.598523    0.0     1.3   
六個月內住院次數                6339.0     2.056791    9.258256    0.0     0.0   
first_has_denture       5852.0     0.140294    0.347321    0.0     0.0   
last_has_denture        5852.0     0.122522    0.327916    0.0     0.0   
diff_has_denture        5852.0    -0.017772    0.183097   -1.0     0.0   
first_ 意識_E             2812.0     3.865932    0.484721    1.0     4.0   
last_ 意識_E              2812.0     3.880156    0.449757    1.0     4.0   
first_ 意識_V             2764.0     4.150868    1.167494    1.0     4.0   
last_ 意識_V              2764.0     4.073082    1.182457    1.0     4.0   
first_ 意識_M             2816.0     5.271307    1.134745    1.0     5.0   
last_ 意識_M              2816.0     5.237571    1.130564    1.0     5.0   
first_ 意識總分             2816.0    13.222301    2.406389    3.0    12.0   
last_ 意識總分              2816.0    13.126420    2.385048    3.0    12.0   
意識總分Max                 5852.0     8.005297    5.428563    3.0     3.0   
意識總分_diff               2757.0    -0.099383    1.168150  -11.0     0.0   
意識分級                       0.0          NaN         NaN    NaN     NaN   
使用呼吸輔具                  5852.0     0.081681    0.273902    0.0     0.0   
first_has_feeding_tube  5852.0     0.153623    0.360618    0.0     0.0   
last_has_feeding_tube   5852.0     0.173103    0.378369    0.0     0.0   
diff_has_feeding_tube   5852.0     0.019481    0.198171   -1.0     0.0   
had_fall                5852.0     0.115516    0.319671    0.0     0.0   
呼吸_ADL低                 6339.0     0.061682    0.240595    0.0     0.0   
高齡_住院多                  6339.0     0.061208    0.239731    0.0     0.0   
ADL低_明顯惡化               6339.0     0.030604    0.172256    0.0     0.0   
DNR_呼吸                  6339.0     0.049850    0.217652    0.0     0.0   
超級重症                    6339.0     0.017826    0.132330    0.0     0.0   
low_adl_no_var          6339.0     0.103644    0.304822    0.0     0.0   

                           50%          75%           max  
H01_NUM                 1265.0  1386.000000  20295.000000  
dbname                     NaN          NaN           Na

In [11]:
ex_missing_info = external.isnull().sum().to_frame(name='Missing Count')
ex_missing_info['Missing Ratio'] = (ex_missing_info['Missing Count'] / len(external)).round(4)
ex_missing_info = ex_missing_info.sort_values(by='Missing Ratio', ascending=True)
ex_missing_info

Missing Count  Missing Ratio
H01_NUM                             0         0.0000
性別_is_male                          0         0.0000
觀察天數                                0         0.0000
死亡標記                                0         0.0000
預估年齡                                0         0.0000
ADL_Max                             0         0.0000
ADL_總分_max                          0         0.0000
DNR_flag                            0         0.0000
ADL_Min                             0         0.0000
ADL_last_CouldNot                   0         0.0000
ADL_first_CouldNot                  0         0.0000
ADL_明顯惡化                            0         0.0000
六個月內住院次數                            0         0.0000
高齡_住院多                              0         0.0000
呼吸_ADL低                             0         0.0000
ADL低_明顯惡化                           0         0.0000
DNR_呼吸                              0         0.0000
超級重症                                0         0.0000
low_adl_no_var                      0         0.0000
ADL_first_score                    66         0.0104
ADL_last_score                     67         0.0106
ADL_diff_seq                       93         0.0147
diff_has_denture                  487         0.0768
last_has_denture                  487         0.0768
意識總分Max                           487         0.0768
使用呼吸輔具                            487         0.0768
first_has_denture                 487         0.0768
had_fall                          487         0.0768
diff_has_feeding_tube             487         0.0768
last_has_feeding_tube             487         0.0768
first_has_feeding_tube            487         0.0768
ADL_std                          1505         0.2374
last_ 意識總分                       3523         0.5558
last_ 意識_M                       3523         0.5558
first_ 意識_M                      3523         0.5558
first_ 意識總分                      3523         0.5558
last_ 意識_E                       3527         0.5564
first_ 意識_E                      3527         0.5564
last_ 意識_V                       3575         0.5640
first_ 意識_V                      3575         0.5640
意識總分_diff                        3582         0.5651
disease_Max                      3819         0.6025
disease_Min                      3819         0.6025
disease_last_score               3833         0.6047
disease_first_score              3836         0.6051
disease_diff_seq                 3838         0.6055
disease_std                      4671         0.7369
dbname                           6339         1.0000
結案日期                             6339         1.0000
入家日期                             6339         1.0000
意識分級                             6339         1.0000

訓練資料讀入

In [12]:
# read data and put it in a dataframe
# 在 google 工作表載入訓練資料 gsheets

gsheets = gc.open_by_url('https://docs.google.com/spreadsheets/d/1_D_ryADEhFBC0322zwlI7uQZz5xELWlZ-3CO549cLS0/edit?usp=sharing')
worksheet = gsheets.worksheet("training_data_v2_0917")  # 指定分頁名稱

worksheet = worksheet.get_all_records()
df = pd.DataFrame(worksheet)
df = df.apply(lambda col: pd.to_numeric(col.astype(str).str.replace(',', '').str.strip(), errors='coerce'))
df.head()

H01_NUM  dbname  入家日期  結案日期  死亡標記  觀察天數  性別_is_male  預估年齡  DNR_flag  \
0     1325     NaN   NaN   NaN     0   739           1    66         1   
1     1160     NaN   NaN   NaN     0   788           1    89         1   
2     1253     NaN   NaN   NaN     0  1292           0    89         0   
3     1342     NaN   NaN   NaN     0   584           0    93         1   
4     1343     NaN   NaN   NaN     0   583           1    90         0   

   ADL_總分_max  ...  first_ 意識總分  last_ 意識總分  意識總分Max  意識總分_diff  意識分級  使用呼吸輔具  \
0           0  ...         13.0        12.0     13.0       -1.0   NaN     0.0   
1          95  ...         15.0        13.0     15.0       -2.0   NaN     0.0   
2          60  ...         15.0        15.0     15.0        0.0   NaN     0.0   
3          30  ...         14.0        14.0     14.0        0.0   NaN     0.0   
4          90  ...         15.0        15.0     15.0        0.0   NaN     0.0   

   first_has_feeding_tube  last_has_feeding_tube  diff_has_feeding_tube  \
0                     0.0                    0.0                    0.0   
1                     0.0                    0.0                    0.0   
2                     0.0                    0.0                    0.0   
3                     0.0                    0.0                    0.0   
4                     0.0                    0.0                    0.0   

   had_fall  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  

[5 rows x 45 columns]

In [13]:
df['呼吸_ADL低'] = ((df['使用呼吸輔具'] == 1) & (df['ADL_總分_max'] < 30)).astype(int)
df['高齡_住院多'] = ((df['預估年齡'] > 85) & (df['六個月內住院次數'] > 2)).astype(int)
df['ADL低_明顯惡化'] = ((df['ADL_總分_max'] < 30) & (df['ADL_明顯惡化'] == 1)).astype(int)
df['DNR_呼吸'] = ((df['DNR_flag'] == 1) & (df['使用呼吸輔具'] == 1)).astype(int)
df['超級重症'] = (
    (df['DNR_flag'] == 1) &
    (df['使用呼吸輔具'] == 1) &
    (df['ADL_總分_max'] < 30) &
    (df['六個月內住院次數'] > 1)
).astype(int)
df["low_adl_no_var"] = ((df["ADL_std"].isnull()) & (df["ADL_總分_max"] < 5)).astype(int)

In [14]:
df.describe(include='all').T

count         mean          std    min     25%  \
H01_NUM                 27037.0  1500.801457  4160.182209    1.0  1159.0   
dbname                      0.0          NaN          NaN    NaN     NaN   
入家日期                        0.0          NaN          NaN    NaN     NaN   
結案日期                        0.0          NaN          NaN    NaN     NaN   
死亡標記                    27037.0     0.194992     0.396202    0.0     0.0   
觀察天數                    27037.0   634.100529   437.553151    0.0   248.0   
性別_is_male              27037.0     0.485557     0.499801    0.0     0.0   
預估年齡                    27037.0    79.388801    11.819412    0.0    72.0   
DNR_flag                27037.0     0.408810     0.491623    0.0     0.0   
ADL_總分_max              27037.0    31.974886    33.277403    0.0     0.0   
ADL_first_score         26567.0    28.209245    31.598174    0.0     0.0   
ADL_last_score          26510.0    24.693135    31.141733    0.0     0.0   
ADL_diff_seq            26320.0    -3.420213    19.495579 -100.0    -5.0   
ADL_std                 22796.0     5.900205     9.049853    0.0     0.0   
ADL_Max                 27037.0    31.974886    33.277403    0.0     0.0   
ADL_Min                 27037.0    20.445131    28.644584    0.0     0.0   
ADL_明顯惡化                27037.0     0.194955     0.396173    0.0     0.0   
ADL_first_CouldNot      27037.0     0.017384     0.130698    0.0     0.0   
ADL_last_CouldNot       27037.0     0.019492     0.138248    0.0     0.0   
disease_first_score      7584.0     1.673536     0.579513    0.0     1.3   
disease_last_score       7610.0     1.683338     0.539606    0.0     1.3   
disease_diff_seq         7550.0     0.011298     0.346996   -3.0     0.0   
disease_std              5888.0     0.115417     0.209287    0.0     0.0   
disease_Max              7704.0     1.749221     0.581545    0.0     1.4   
disease_Min              7704.0     1.557282     0.595950    0.0     1.2   
六個月內住院次數                27037.0     2.779672    10.889087    0.0     0.0   
first_has_denture       24995.0     0.186277     0.389338    0.0     0.0   
last_has_denture        24995.0     0.157832     0.364590    0.0     0.0   
diff_has_denture        24995.0    -0.028446     0.251170   -1.0     0.0   
first_ 意識_E             13022.0     3.853018     0.486566    1.0     4.0   
last_ 意識_E              13022.0     3.849946     0.485944    1.0     4.0   
first_ 意識_V             12724.0     4.042361     1.243143    1.0     4.0   
last_ 意識_V              12724.0     3.918186     1.275093    1.0     3.0   
first_ 意識_M             13024.0     5.205620     1.164545    1.0     5.0   
last_ 意識_M              13024.0     5.124386     1.185430    1.0     5.0   
first_ 意識總分             13024.0    13.025184     2.507728    3.0    12.0   
last_ 意識總分              13024.0    12.819564     2.555053    3.0    11.0   
意識總分Max                 24995.0     8.338508     5.393691    3.0     3.0   
意識總分_diff               12701.0    -0.212660     1.358108  -12.0     0.0   
意識分級                        0.0          NaN          NaN    NaN     NaN   
使用呼吸輔具                  24995.0     0.055171     0.228318    0.0     0.0   
first_has_feeding_tube  24995.0     0.086297     0.280808    0.0     0.0   
last_has_feeding_tube   24995.0     0.130746     0.337129    0.0     0.0   
diff_has_feeding_tube   24995.0     0.044449     0.237146   -1.0     0.0   
had_fall                24995.0     0.065253     0.246977    0.0     0.0   
呼吸_ADL低                 27037.0     0.042201     0.201052    0.0     0.0   
高齡_住院多                  27037.0     0.099271     0.299032    0.0     0.0   
ADL低_明顯惡化               27037.0     0.051411     0.220839    0.0     0.0   
DNR_呼吸                  27037.0     0.030329     0.171494    0.0     0.0   
超級重症                    27037.0     0.013574     0.115716    0.0     0.0   
low_adl_no_var          27037.0     0.072900     0.259977    0.0     0.0   

                                50%          75%            max  
H01_NUM

In [15]:
df[df['六個月內住院次數'] == 618]

H01_NUM  dbname  入家日期  結案日期  死亡標記  觀察天數  性別_is_male  預估年齡  DNR_flag  \
27035     1190     NaN   NaN   NaN     0  1385           0    54         0   

       ADL_總分_max  ...  first_has_feeding_tube  last_has_feeding_tube  \
27035         100  ...                     0.0                    0.0   

       diff_has_feeding_tube  had_fall  呼吸_ADL低  高齡_住院多  ADL低_明顯惡化  DNR_呼吸  \
27035                    0.0       0.0        0       0          0       0   

       超級重症  low_adl_no_var  
27035     0               0  

[1 rows x 51 columns]

In [22]:
# 顯示所有欄位
pd.set_option('display.max_columns', None)

# 顯示所有列
pd.set_option('display.max_rows', None)

# 顯示寬度不要自動換行
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

In [23]:
df[(df['ADL_diff_seq'].isnull()) &
   (df['死亡標記'] == 1) &
   (df['ADL_std'].notnull())]

H01_NUM  dbname  入家日期  結案日期  死亡標記  觀察天數  性別_is_male  預估年齡  DNR_flag  \
11643     1338     NaN   NaN   NaN     1   154           1    67         0   
11686     1338     NaN   NaN   NaN     1   133           1    92         0   
11735     1236     NaN   NaN   NaN     1    32           0    70         1   
11808     1307     NaN   NaN   NaN     1    78           1    87         0   
11884     1295     NaN   NaN   NaN     1   135           1    84         0   
11886     1444     NaN   NaN   NaN     1    34           1    74         0   
12092     1342     NaN   NaN   NaN     1   147           1    87         1   
12099     1270     NaN   NaN   NaN     1    75           1    73         1   
12101     1269     NaN   NaN   NaN     1   108           0    79         0   
12102     1274     NaN   NaN   NaN     1    91           1    84         1   
12108     1279     NaN   NaN   NaN     1   120           1    84         1   
12111     1283     NaN   NaN   NaN     1    64           1    85         1   
12233     1212     NaN   NaN   NaN     1    48           1    72         1   
12243     1433     NaN   NaN   NaN     1    27           1    86         0   
12263     1298     NaN   NaN   NaN     1    85           1    62         0   
12315     1274     NaN   NaN   NaN     1    88           1    81         0   
12579     1473     NaN   NaN   NaN     1   111           0    60         1   
12666     1354     NaN   NaN   NaN     1    93           1    86         0   
12708     5265     NaN   NaN   NaN     1    80           1    82         0   
12743     1273     NaN   NaN   NaN     1    20           0    89         1   
12806     1393     NaN   NaN   NaN     1   165           0    85         0   
12878     1260     NaN   NaN   NaN     1    52           0    86         1   
12899     1760     NaN   NaN   NaN     1    34           1    77         1   
13228     3673     NaN   NaN   NaN     1    78           0   100         1   
13252     1329     NaN   NaN   NaN     1    97           1    81         1   
13271       31     NaN   NaN   NaN     1     0           0    88         0   
13301     1480     NaN   NaN   NaN     1    44           0    92         1   
13385     1337     NaN   NaN   NaN     1   147           0    88         1   
13563     1322     NaN   NaN   NaN     1   144           0    83         0   
13620     1690     NaN   NaN   NaN     1    38           1    67         0   
13715     1329     NaN   NaN   NaN     1    66           1    95         1   
13824     1339     NaN   NaN   NaN     1    37           1    63         1   
13835     1311     NaN   NaN   NaN     1   117           1    87         1   
13848     1277     NaN   NaN   NaN     1    83           1    77         1   
13902     1210     NaN   NaN   NaN     1   114           0    92         1   
13914     1314     NaN   NaN   NaN     1    25           1    68         1   
14035     1188     NaN   NaN   NaN     1    60           1    88         1   
14062     1292     NaN   NaN   NaN     1    68           1    86         1   
14231     1183     NaN   NaN   NaN     1   100           0    80         0   
14337     1326     NaN   NaN   NaN     1    34           0    95         0   
14460     1221     NaN   NaN   NaN     1   110           0    90         0   
14521     1223     NaN   NaN   NaN     1   178           1    92         1   
14523     1210     NaN   NaN   NaN     1   107           0    86         1   
14584     1311     NaN   NaN   NaN     1   109           1    76         1   
14622     1237     NaN   NaN   NaN     1   111           0    91         1   
14623     1186     NaN   NaN   NaN     1    69           1    81         1   
14629     1124     NaN   NaN   NaN     1   167           0    83         0   
14630     1289     NaN   NaN   NaN     1   108           1    51         1   
14701     1210     NaN   NaN   NaN     1   143           0    82         0   
14704     1197     NaN   NaN   NaN     1    88           1    96         1   
14783     1134     NaN   NaN 

In [24]:
len(df[(df['ADL_diff_seq'].isnull()) &
   (df['死亡標記'] == 1) &
   (df['ADL_std'].notnull())])

99

In [ ]:
df_missing_info = df.isnull().sum().to_frame(name='Missing Count')
df_missing_info['Missing Ratio'] = (df_missing_info['Missing Count'] / len(df)).round(4)
df_missing_info = df_missing_info.sort_values(by='Missing Ratio', ascending=True)
df_missing_info

Missing Count  Missing Ratio
H01_NUM                             0         0.0000
性別_is_male                          0         0.0000
觀察天數                                0         0.0000
死亡標記                                0         0.0000
預估年齡                                0         0.0000
ADL_Max                             0         0.0000
ADL_總分_max                          0         0.0000
DNR_flag                            0         0.0000
ADL_Min                             0         0.0000
ADL_last_CouldNot                   0         0.0000
ADL_first_CouldNot                  0         0.0000
ADL_明顯惡化                            0         0.0000
六個月內住院次數                            0         0.0000
高齡_住院多                              0         0.0000
呼吸_ADL低                             0         0.0000
ADL低_明顯惡化                           0         0.0000
DNR_呼吸                              0         0.0000
超級重症                                0         0.0000
low_adl_no_var                      0         0.0000
ADL_first_score                   470         0.0174
ADL_last_score                    527         0.0195
ADL_diff_seq                      717         0.0265
diff_has_denture                 2042         0.0755
last_has_denture                 2042         0.0755
意識總分Max                          2042         0.0755
使用呼吸輔具                           2042         0.0755
first_has_denture                2042         0.0755
had_fall                         2042         0.0755
diff_has_feeding_tube            2042         0.0755
last_has_feeding_tube            2042         0.0755
first_has_feeding_tube           2042         0.0755
ADL_std                          4241         0.1569
last_ 意識總分                      14013         0.5183
last_ 意識_M                      14013         0.5183
first_ 意識_M                     14013         0.5183
first_ 意識總分                     14013         0.5183
last_ 意識_E                      14015         0.5184
first_ 意識_E                     14015         0.5184
last_ 意識_V                      14313         0.5294
first_ 意識_V                     14313         0.5294
意識總分_diff                       14336         0.5302
disease_Max                     19333         0.7151
disease_Min                     19333         0.7151
disease_last_score              19427         0.7185
disease_first_score             19453         0.7195
disease_diff_seq                19487         0.7208
disease_std                     21149         0.7822
dbname                          27037         1.0000
結案日期                            27037         1.0000
入家日期                            27037         1.0000
意識分級                            27037         1.0000

In [ ]:
features = df_missing_info[df_missing_info['Missing Ratio']<0.3].index.tolist()

In [ ]:
features

['H01_NUM',
 '性別_is_male',
 '觀察天數',
 '死亡標記',
 '預估年齡',
 'ADL_Max',
 'ADL_總分_max',
 'DNR_flag',
 'ADL_Min',
 'ADL_last_CouldNot',
 'ADL_first_CouldNot',
 'ADL_明顯惡化',
 '六個月內住院次數',
 '高齡_住院多',
 '呼吸_ADL低',
 'ADL低_明顯惡化',
 'DNR_呼吸',
 '超級重症',
 'low_adl_no_var',
 'ADL_first_score',
 'ADL_last_score',
 'ADL_diff_seq',
 'diff_has_denture',
 'last_has_denture',
 '意識總分Max',
 '使用呼吸輔具',
 'first_has_denture',
 'had_fall',
 'diff_has_feeding_tube',
 'last_has_feeding_tube',
 'first_has_feeding_tube',
 'ADL_std']

In [ ]:
dfNew = df[features]

In [ ]:
dfNew = dfNew.fillna(0)

In [ ]:
from sklearn.base import BaseEstimator, ClassifierMixin

class HybridXGBRF(BaseEstimator, ClassifierMixin):
    def __init__(self, xgb_model=None, rf_model=None, alpha=0.5):
        self.xgb_model = xgb_model
        self.rf_model = rf_model
        self.alpha = alpha
        self._init_models()

    def _init_models(self):
        # Best Parameters: {'colsample_bytree': 0.8, 'learning_rate': 0.01, 'max_depth': 4, 'n_estimators': 800, 'subsample': 1.0}
        # "XGBClassifier": XGBClassifier(n_estimators=200, learning_rate=0.01, max_depth=5, random_state=42, eval_metric='logloss'),

        self.xgb = self.xgb_model or XGBClassifier(
            eval_metric="logloss",
            random_state=42,
            colsample_bytree=0.8,     # ✅ 降低每棵樹看到的特徵比例 → 提高多樣性
            learning_rate=0.01,       # ✅ 稍微提升學習率搭配更早停止
            max_depth=5,              # ✅ 降低單棵樹複雜度 → 降低過擬合
            n_estimators=200,         # ✅ 總樹數可略減以免累積錯誤
            subsample=1.0,            # ✅ 樣本隨機抽樣 → 提升隨機性
            verbosity=0,
            use_label_encoder=False
        )
        self.rf = self.rf_model or RandomForestClassifier(
            n_estimators=100,
            random_state=42
        )

    def fit(self, X, y):
        self._init_models()  # 每次 fit 要重設模型
        self.xgb.fit(X, y)
        self.rf.fit(X, y)
        return self

    def predict_proba(self, X):
        xgb_prob = self.xgb.predict_proba(X)[:, 1]
        rf_prob = self.rf.predict_proba(X)[:, 1]
        blended = self.alpha * xgb_prob + (1 - self.alpha) * rf_prob
        return np.vstack([1 - blended, blended]).T

    def predict(self, X):
        return (self.predict_proba(X)[:, 1] > 0.5).astype(int)

    def get_params(self, deep=True):
        return {
            'xgb_model': self.xgb_model,
            'rf_model': self.rf_model,
            'alpha': self.alpha
        }

    def set_params(self, **params):
        for param, value in params.items():
            setattr(self, param, value)
        self._init_models()  # 重新初始化模型
        return self

In [ ]:
# 要移除的欄位，是代表身分標記，以及天數
drop_columns = ['H01_NUM', '觀察天數']

# 丟掉這些欄位
dfNew = dfNew.drop(columns=drop_columns)

In [ ]:
dfNew

性別_is_male  死亡標記  預估年齡  ADL_Max  ADL_總分_max  DNR_flag  ADL_Min  \
0               1     0    66        0           0         1        0   
1               1     0    89       95          95         1       10   
2               0     0    89       60          60         0       30   
3               0     0    93       30          30         1       10   
4               1     0    90       90          90         0        5   
...           ...   ...   ...      ...         ...       ...      ...   
27032           1     0    48      100         100         0      100   
27033           1     0    41      100         100         0      100   
27034           0     0    60      100         100         0      100   
27035           0     0    54      100         100         0      100   
27036           1     0    58      100         100         0      100   

       ADL_last_CouldNot  ADL_first_CouldNot  ADL_明顯惡化  ...  diff_has_denture  \
0                      0                   0         0  ...               0.0   
1                      0                   0         1  ...              -1.0   
2                      0                   0         1  ...               0.0   
3                      0                   0         1  ...               0.0   
4                      0                   0         1  ...               0.0   
...                  ...                 ...       ...  ...               ...   
27032                  0                   0         0  ...               0.0   
27033                  0                   0         0  ...               0.0   
27034                  0                   0         0  ...               0.0   
27035                  0                   0         0  ...               0.0   
27036                  0                   0         0  ...               0.0   

       last_has_denture  意識總分Max  使用呼吸輔具  first_has_denture  had_fall  \
0                   0.0     13.0     0.0                0.0       0.0   
1                   0.0     15.0     0.0                1.0       0.0   
2                   0.0     15.0     0.0                0.0       0.0   
3                   0.0     14.0     0.0                0.0       0.0   
4                   0.0     15.0     0.0                0.0       0.0   
...                 ...      ...     ...                ...       ...   
27032               0.0      3.0     0.0                0.0       0.0   
27033               0.0      3.0     0.0                0.0       0.0   
27034               0.0      3.0     0.0                0.0       0.0   
27035               0.0      3.0     0.0                0.0       0.0   
27036               1.0      3.0     0.0                1.0       0.0   

       diff_has_feeding_tube  last_has_feeding_tube  first_has_feeding_tube  \
0                        0.0                    0.0                     0.0   
1                        0.0                    0.0                     0.0   
2                        0.0                    0.0                     0.0   
3                        0.0                    0.0                     0.0   
4                        0.0                    0.0                     0.0   
...                      ...                    ...                     ...   
27032                    0.0                    0.0                     0.0   
27033                    0.0                    0.0                     0.0   
27034                    0.0                    0.0                     0.0   
27035                    0.0                    0.0                     0.0   
27036                    0.0                    0.0                     0.0   

         ADL_std  
0       0.000000  
1      49.074773  
2      16.431677  
3      10.350983  
4      45.434411  
...          ...  
27032   0.000000  
27033   0.000000  
27034   0.000000  
27035   0.000000  
27036   0.000000  

[27037 rows x 30 columns]

In [ ]:
import numpy as np
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold
from sklearn.metrics import (
    confusion_matrix, roc_curve, auc, classification_report
)
import plotly.graph_objects as go
import plotly.express as px

In [ ]:
# === 資料準備 ===
X = dfNew.drop(columns=['死亡標記'])
y = df['死亡標記']

In [ ]:
X_missing_info = X.isnull().sum().to_frame(name='Missing Count')
X_missing_info['Missing Ratio'] = (X_missing_info['Missing Count'] / len(X)).round(4)
X_missing_info = X_missing_info.sort_values(by='Missing Ratio', ascending=True)
X_missing_info

Missing Count  Missing Ratio
性別_is_male                          0            0.0
預估年齡                                0            0.0
ADL_Max                             0            0.0
ADL_總分_max                          0            0.0
DNR_flag                            0            0.0
ADL_Min                             0            0.0
ADL_last_CouldNot                   0            0.0
ADL_first_CouldNot                  0            0.0
ADL_明顯惡化                            0            0.0
六個月內住院次數                            0            0.0
高齡_住院多                              0            0.0
呼吸_ADL低                             0            0.0
ADL低_明顯惡化                           0            0.0
DNR_呼吸                              0            0.0
超級重症                                0            0.0
low_adl_no_var                      0            0.0
ADL_first_score                     0            0.0
ADL_last_score                      0            0.0
ADL_diff_seq                        0            0.0
diff_has_denture                    0            0.0
last_has_denture                    0            0.0
意識總分Max                             0            0.0
使用呼吸輔具                              0            0.0
first_has_denture                   0            0.0
had_fall                            0            0.0
diff_has_feeding_tube               0            0.0
last_has_feeding_tube               0            0.0
first_has_feeding_tube              0            0.0
ADL_std                             0            0.0

# 開始進行訓練

In [ ]:
X.describe(include='all').T

count       mean        std    min   25%   50%  \
性別_is_male              27037.0   0.485557   0.499801    0.0   0.0   0.0   
預估年齡                    27037.0  79.388801  11.819412    0.0  72.0  82.0   
ADL_Max                 27037.0  31.974886  33.277403    0.0   0.0  20.0   
ADL_總分_max              27037.0  31.974886  33.277403    0.0   0.0  20.0   
DNR_flag                27037.0   0.408810   0.491623    0.0   0.0   0.0   
ADL_Min                 27037.0  20.445131  28.644584    0.0   0.0   5.0   
ADL_last_CouldNot       27037.0   0.019492   0.138248    0.0   0.0   0.0   
ADL_first_CouldNot      27037.0   0.017384   0.130698    0.0   0.0   0.0   
ADL_明顯惡化                27037.0   0.194955   0.396173    0.0   0.0   0.0   
六個月內住院次數                27037.0   2.779672  10.889087    0.0   0.0   1.0   
高齡_住院多                  27037.0   0.099271   0.299032    0.0   0.0   0.0   
呼吸_ADL低                 27037.0   0.042201   0.201052    0.0   0.0   0.0   
ADL低_明顯惡化               27037.0   0.051411   0.220839    0.0   0.0   0.0   
DNR_呼吸                  27037.0   0.030329   0.171494    0.0   0.0   0.0   
超級重症                    27037.0   0.013574   0.115716    0.0   0.0   0.0   
low_adl_no_var          27037.0   0.072900   0.259977    0.0   0.0   0.0   
ADL_first_score         27037.0  27.718867  31.538558    0.0   0.0  15.0   
ADL_last_score          27037.0  24.211821  31.025110    0.0   0.0  10.0   
ADL_diff_seq            27037.0  -3.329511  19.243177 -100.0  -5.0   0.0   
diff_has_denture        27037.0  -0.026297   0.241615   -1.0   0.0   0.0   
last_has_denture        27037.0   0.145911   0.353024    0.0   0.0   0.0   
意識總分Max                 27037.0   7.708732   5.634675    0.0   3.0   3.0   
使用呼吸輔具                  27037.0   0.051004   0.220010    0.0   0.0   0.0   
first_has_denture       27037.0   0.172208   0.377569    0.0   0.0   0.0   
had_fall                27037.0   0.060325   0.238092    0.0   0.0   0.0   
diff_has_feeding_tube   27037.0   0.041092   0.228317   -1.0   0.0   0.0   
last_has_feeding_tube   27037.0   0.120871   0.325984    0.0   0.0   0.0   
first_has_feeding_tube  27037.0   0.079780   0.270957    0.0   0.0   0.0   
ADL_std                 27037.0   4.974704   8.582360    0.0   0.0   0.0   

                              75%         max  
性別_is_male               1.000000    1.000000  
預估年齡                    88.000000  125.000000  
ADL_Max                 55.000000  100.000000  
ADL_總分_max              55.000000  100.000000  
DNR_flag                 1.000000    1.000000  
ADL_Min                 30.000000  100.000000  
ADL_last_CouldNot        0.000000    1.000000  
ADL_first_CouldNot       0.000000    1.000000  
ADL_明顯惡化                 0.000000    1.000000  
六個月內住院次數                 3.000000  618.000000  
高齡_住院多                   0.000000    1.000000  
呼吸_ADL低                  0.000000    1.000000  
ADL低_明顯惡化                0.000000    1.000000  
DNR_呼吸                   0.000000    1.000000  
超級重症                     0.000000    1.000000  
low_adl_no_var           0.000000    1.000000  
ADL_first_score         50.000000  100.000000  
ADL_last_score          40.000000  100.000000  
ADL_diff_seq             0.000000  100.000000  
diff_has_denture         0.000000    1.000000  
last_has_denture         0.000000    1.000000  
意識總分Max                 14.000000   15.000000  
使用呼吸輔具                   0.000000    1.000000  
first_has_denture        0.000000    1.000000  
had_fall                 0.000000    1.000000  
diff_has_feeding_tube    0.000000    1.000000  
last_has_feeding_tube    0.000000    1.000000  
first_has_feeding_tube   0.000000    1.000000  
ADL_std                  6.614378   70.710678

In [ ]:
!pip install lifelines

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 349.3/349.3 kB 13.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 117.3/117.3 kB 11.6 MB/s eta 0:00:00
  Created wheel for autograd-gamma: filename=autograd_gamma-0.5.0-py3-none-any.whl size=4030 sha256=9af7b202a64eb42134710d8245f517cc8613dd45acd31e4ca4aecc9a29765aaf
  Stored in directory: /root/.cache/pip/wheels/50/37/21/0a719b9d89c635e89ff24bd93b862882ad675279552013b2fb
Successfully built autograd-gamma


In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from lifelines import CoxPHFitter
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegressionCV

# 假設 HybridXGBRF 已定義
all_models = {
    #"HybridXGBRF (Our Approach)": HybridXGBRF(alpha=1),
    "HybridXGBRF (Our Approach)": XGBClassifier(n_estimators=200, learning_rate=0.01, max_depth=5, random_state=42, eval_metric='logloss', subsample=1.0, verbosity=0),
    "LogisticRegression (max_iter=200)": LogisticRegression(max_iter=200),
    "XGBClassifier": XGBClassifier(n_estimators=500, learning_rate=0.01, max_depth=3, random_state=42, eval_metric='logloss'),
    "RandomForestClassifier": RandomForestClassifier(n_estimators=100, random_state=42),
    "LogisticRegression (max_iter=1000)": LogisticRegression(max_iter=1000),

    # 🔽 新增未測試模型
    "Ridge": make_pipeline(StandardScaler(), LogisticRegression(penalty='l2', solver='saga', max_iter=1000, random_state=42)),
    "Lasso": make_pipeline(StandardScaler(), LogisticRegression(penalty='l1', solver='saga', max_iter=1000, random_state=42)),
    "Elastic": make_pipeline(StandardScaler(), LogisticRegression(penalty='elasticnet', solver='saga', l1_ratio=0.5, max_iter=1000, random_state=42)),
}

In [ ]:
import copy
from sklearn.base import clone
from sklearn.impute import SimpleImputer
from sklearn.exceptions import NotFittedError
from sklearn.utils.validation import check_is_fitted

# Prepare CV and ROC Figure
n_splits = 5
skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=42)
fig_roc = go.Figure()
mean_fpr = np.linspace(0, 1, 100)

results = []
trained_models = {}

for model_name, model in all_models.items():
    print(f"▶ Running CV for: {model_name}")
    accs, precs, recalls, f1s, aucs = [], [], [], [], []
    all_cm = np.zeros((2, 2), dtype=int)
    tprs = []

    for fold, (train_idx, test_idx) in enumerate(skf.split(X, y)):
        X_train, X_test = X.iloc[train_idx].copy(), X.iloc[test_idx].copy()
        y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]

        model_fold = clone(model)

        try:
            # 嘗試直接 fit（若模型支援 NaN，會成功）
            model_fold.fit(X_train, y_train)

        except ValueError as e:
            if "Input X contains NaN" in str(e):
                print(f"⚠️ Missing value detected for {model_name} (fold {fold+1}) — applying median imputation.")

                imputer = SimpleImputer(strategy='median')
                X_train = pd.DataFrame(imputer.fit_transform(X_train), columns=X.columns)
                X_test = pd.DataFrame(imputer.transform(X_test), columns=X.columns)

                model_fold.fit(X_train, y_train)
            else:
                raise e  # 若是其他錯誤就直接拋出

        y_pred = model_fold.predict(X_test)
        y_prob = model_fold.predict_proba(X_test)[:, 1]

        # ROC
        fpr, tpr, _ = roc_curve(y_test, y_prob)
        roc_auc = auc(fpr, tpr)
        tpr_interp = np.interp(mean_fpr, fpr, tpr)
        tpr_interp[0] = 0.0
        tprs.append(tpr_interp)
        aucs.append(roc_auc)

        # Metrics
        accs.append(accuracy_score(y_test, y_pred))
        precs.append(precision_score(y_test, y_pred))
        recalls.append(recall_score(y_test, y_pred))
        f1s.append(f1_score(y_test, y_pred))
        all_cm += confusion_matrix(y_test, y_pred)

    trained_models[model_name] = copy.deepcopy(model_fold)

    # ROC Curve
    mean_tpr = np.mean(tprs, axis=0)
    mean_tpr[-1] = 1.0
    mean_auc = auc(mean_fpr, mean_tpr)

    fig_roc.add_trace(go.Scatter(
        x=mean_fpr, y=mean_tpr, mode='lines',
        name=f"{model_name} (mean AUC={mean_auc:.3f})"
    ))

    results.append({
        'Model': model_name,
        'Accuracy Mean': np.mean(accs),
        'Accuracy Std': np.std(accs),
        'Precision Mean': np.mean(precs),
        'Recall Mean': np.mean(recalls),
        'F1 Score Mean': np.mean(f1s),
        'ROC AUC Mean': np.mean(aucs),
        'ROC AUC Std': np.std(aucs),
        'Confusion Matrix': all_cm
    })

# Add Random Baseline
fig_roc.add_trace(go.Scatter(
    x=[0, 1], y=[0, 1], mode='lines',
    line=dict(dash='dash'), name='Random Baseline'
))

fig_roc.update_layout(
    title="ROC Curve Comparison (Cross-Validation Mean)",
    xaxis_title="False Positive Rate",
    yaxis_title="True Positive Rate",
    width=800, height=600
)

fig_roc.show()

▶ Running CV for: HybridXGBRF (Our Approach)
▶ Running CV for: LogisticRegression (max_iter=200)


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

/usr/local/lib/python3.12/dist-packages/sklearn/

▶ Running CV for: XGBClassifier
▶ Running CV for: RandomForestClassifier
▶ Running CV for: LogisticRegression (max_iter=1000)


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

/usr/local/lib/python3.12/dist-packages/sklearn/

▶ Running CV for: Ridge
▶ Running CV for: Lasso


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning:

The max_iter was reached which means the coef_ did not converge

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning:

The max_iter was reached which means the coef_ did not converge

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning:

The max_iter was reached which means the coef_ did not converge

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning:

The max_iter was reached which means the coef_ did not converge

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning:

The max_iter was reached which means the coef_ did not converge



▶ Running CV for: Elastic


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning:

The max_iter was reached which means the coef_ did not converge

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning:

The max_iter was reached which means the coef_ did not converge

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning:

The max_iter was reached which means the coef_ did not converge



In [ ]:
import shap

# 訓練模型（完整資料）
xgb_model = all_models["HybridXGBRF (Our Approach)"]
xgb_model.fit(X, y)

# 建立 SHAP 解釋器
explainer = shap.TreeExplainer(xgb_model)
shap_values = explainer.shap_values(X)

# 計算平均 SHAP 值絕對值（作為重要性）
import numpy as np
shap_abs_mean = np.abs(shap_values).mean(axis=0)

In [ ]:
import pandas as pd
import plotly.express as px

# 整理成 DataFrame 並排序
importance_df = pd.DataFrame({
    'Feature': X.columns,
    'Mean |SHAP Value|': shap_abs_mean
}).sort_values(by='Mean |SHAP Value|', ascending=False)

# 畫前 20 名
top_n = 20
fig_bar = px.bar(
    importance_df.head(top_n),
    x='Mean |SHAP Value|',
    y='Feature',
    orientation='h',
    title="🎯 Top SHAP Features by Mean |SHAP|",
)
fig_bar.update_layout(yaxis=dict(categoryorder='total ascending'))
fig_bar.show()

In [ ]:
import plotly.graph_objects as go

for r in results:
    cm = r['Confusion Matrix']
    model_name = r['Model']

    z = cm
    x_labels = ['Predicted 0', 'Predicted 1']
    y_labels = ['Actual 0', 'Actual 1']

    fig_cm = go.Figure(data=go.Heatmap(
        z=z,
        x=x_labels,
        y=y_labels,
        colorscale='Blues',
        text=z,
        texttemplate="%{text}"
    ))

    fig_cm.update_layout(
        title=f"Confusion Matrix - {model_name}",
        xaxis_title="Predicted Label",
        yaxis_title="True Label",
        width=500,
        height=500
    )

    fig_cm.show()

In [ ]:
# Convert results to DataFrame
results_df = pd.DataFrame(results)
results_df = results_df.sort_values(by="ROC AUC Mean", ascending=False).reset_index(drop=True)
results_df

Model  Accuracy Mean  Accuracy Std  \
0          HybridXGBRF (Our Approach)       0.847357      0.000665   
1                       XGBClassifier       0.847690      0.000542   
2  LogisticRegression (max_iter=1000)       0.847838      0.001462   
3                               Lasso       0.847912      0.001484   
4                             Elastic       0.847764      0.001424   
5                               Ridge       0.847727      0.001325   
6   LogisticRegression (max_iter=200)       0.847616      0.001301   
7              RandomForestClassifier       0.831268      0.002996   

   Precision Mean  Recall Mean  F1 Score Mean  ROC AUC Mean  ROC AUC Std  \
0        0.806747     0.286042       0.422123      0.849763     0.003457   
1        0.798973     0.293060       0.428570      0.849151     0.003564   
2        0.778424     0.307476       0.440576      0.820968     0.002809   
3        0.780057     0.306906       0.440237      0.819930     0.002739   
4        0.780206     0.305768       0.439080      0.818999     0.002647   
5        0.780908     0.305009       0.438398      0.818110     0.002574   
6        0.778187     0.306147       0.439154      0.817686     0.002825   
7        0.593051     0.429441       0.498063      0.814686     0.003054   

                Confusion Matrix  
0   [[21402, 363], [3764, 1508]]  
1   [[21374, 391], [3727, 1545]]  
2   [[21302, 463], [3651, 1621]]  
3   [[21307, 458], [3654, 1618]]  
4   [[21309, 456], [3660, 1612]]  
5   [[21312, 453], [3664, 1608]]  
6   [[21303, 462], [3658, 1614]]  
7  [[20211, 1554], [3008, 2264]]

# 測試外部資料在 XGBoost 模型下的結果

In [ ]:
external.head()

H01_NUM  dbname  入家日期  結案日期  死亡標記  觀察天數  性別_is_male  預估年齡  DNR_flag  \
0     1376     NaN   NaN   NaN     0   197           1    77         0   
1     1322     NaN   NaN   NaN     0   327           1    92         0   
2     1319     NaN   NaN   NaN     0   255           1    78         1   
3     1333     NaN   NaN   NaN     0   293           1    82         1   
4     1452     NaN   NaN   NaN     0   341           0    80         1   

   ADL_總分_max  ...  first_has_feeding_tube  last_has_feeding_tube  \
0          90  ...                     0.0                    0.0   
1          10  ...                     0.0                    0.0   
2           5  ...                     0.0                    0.0   
3          10  ...                     0.0                    0.0   
4          10  ...                     0.0                    0.0   

   diff_has_feeding_tube  had_fall  呼吸_ADL低  高齡_住院多  ADL低_明顯惡化  DNR_呼吸  超級重症  \
0                    0.0       0.0        0       0          0       0     0   
1                    0.0       0.0        0       0          0       0     0   
2                    0.0       0.0        0       0          0       0     0   
3                    0.0       0.0        0       0          1       0     0   
4                    0.0       0.0        0       0          0       0     0   

   low_adl_no_var  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 51 columns]

In [ ]:
ex_X = external[features].drop(columns=['死亡標記'])
ex_y = external['死亡標記']

In [ ]:
ex_X = ex_X.fillna(0)

In [ ]:
# 丟掉這些欄位
ex_X = ex_X.drop(columns=drop_columns)

In [ ]:
ex_X.describe().T

count       mean        std    min   25%   50%  \
性別_is_male              6339.0   0.508440   0.499968    0.0   0.0   1.0   
預估年齡                    6339.0  78.607825  11.742108    1.0  72.0  81.0   
ADL_Max                 6339.0  29.695536  31.687105    0.0   0.0  20.0   
ADL_總分_max              6339.0  29.695536  31.687105    0.0   0.0  20.0   
DNR_flag                6339.0   0.456854   0.498174    0.0   0.0   0.0   
ADL_Min                 6339.0  21.894621  28.481462    0.0   0.0  10.0   
ADL_last_CouldNot       6339.0   0.010569   0.102271    0.0   0.0   0.0   
ADL_first_CouldNot      6339.0   0.010412   0.101513    0.0   0.0   0.0   
ADL_明顯惡化                6339.0   0.127149   0.333167    0.0   0.0   0.0   
六個月內住院次數                6339.0   2.056791   9.258256    0.0   0.0   1.0   
高齡_住院多                  6339.0   0.061208   0.239731    0.0   0.0   0.0   
呼吸_ADL低                 6339.0   0.061682   0.240595    0.0   0.0   0.0   
ADL低_明顯惡化               6339.0   0.030604   0.172256    0.0   0.0   0.0   
DNR_呼吸                  6339.0   0.049850   0.217652    0.0   0.0   0.0   
超級重症                    6339.0   0.017826   0.132330    0.0   0.0   0.0   
low_adl_no_var          6339.0   0.103644   0.304822    0.0   0.0   0.0   
ADL_first_score         6339.0  26.220224  30.106670    0.0   0.0  15.0   
ADL_last_score          6339.0  25.325761  30.416023    0.0   0.0  10.0   
ADL_diff_seq            6339.0  -0.885786  16.051802 -100.0   0.0   0.0   
diff_has_denture        6339.0  -0.016406   0.175986   -1.0   0.0   0.0   
last_has_denture        6339.0   0.113109   0.316751    0.0   0.0   0.0   
意識總分Max                 6339.0   7.390282   5.634783    0.0   3.0   3.0   
使用呼吸輔具                  6339.0   0.075406   0.264067    0.0   0.0   0.0   
first_has_denture       6339.0   0.129516   0.335796    0.0   0.0   0.0   
had_fall                6339.0   0.106641   0.308681    0.0   0.0   0.0   
diff_has_feeding_tube   6339.0   0.017984   0.190476   -1.0   0.0   0.0   
last_has_feeding_tube   6339.0   0.159804   0.366453    0.0   0.0   0.0   
first_has_feeding_tube  6339.0   0.141820   0.348893    0.0   0.0   0.0   
ADL_std                 6339.0   3.855188   7.979595    0.0   0.0   0.0   

                              75%         max  
性別_is_male               1.000000    1.000000  
預估年齡                    87.000000  124.000000  
ADL_Max                 50.000000  100.000000  
ADL_總分_max              50.000000  100.000000  
DNR_flag                 1.000000    1.000000  
ADL_Min                 35.000000  100.000000  
ADL_last_CouldNot        0.000000    1.000000  
ADL_first_CouldNot       0.000000    1.000000  
ADL_明顯惡化                 0.000000    1.000000  
六個月內住院次數                 2.000000  581.000000  
高齡_住院多                   0.000000    1.000000  
呼吸_ADL低                  0.000000    1.000000  
ADL低_明顯惡化                0.000000    1.000000  
DNR_呼吸                   0.000000    1.000000  
超級重症                     0.000000    1.000000  
low_adl_no_var           0.000000    1.000000  
ADL_first_score         45.000000  100.000000  
ADL_last_score          45.000000  100.000000  
ADL_diff_seq             0.000000   95.000000  
diff_has_denture         0.000000    1.000000  
last_has_denture         0.000000    1.000000  
意識總分Max                 14.000000   15.000000  
使用呼吸輔具                   0.000000    1.000000  
first_has_denture        0.000000    1.000000  
had_fall                 0.000000    1.000000  
diff_has_feeding_tube    0.000000    1.000000  
last_has_feeding_tube    0.000000    1.000000  
first_has_feeding_tube   0.000000    1.000000  
ADL_std                  4.082483   70.710678

In [ ]:
eX_missing_info = ex_X.isnull().sum().to_frame(name='Missing Count')
eX_missing_info['Missing Ratio'] = (eX_missing_info['Missing Count'] / len(ex_X)).round(4)
eX_missing_info = eX_missing_info.sort_values(by='Missing Ratio', ascending=False)
eX_missing_info

Missing Count  Missing Ratio
性別_is_male                          0            0.0
預估年齡                                0            0.0
ADL_Max                             0            0.0
ADL_總分_max                          0            0.0
DNR_flag                            0            0.0
ADL_Min                             0            0.0
ADL_last_CouldNot                   0            0.0
ADL_first_CouldNot                  0            0.0
ADL_明顯惡化                            0            0.0
六個月內住院次數                            0            0.0
高齡_住院多                              0            0.0
呼吸_ADL低                             0            0.0
ADL低_明顯惡化                           0            0.0
DNR_呼吸                              0            0.0
超級重症                                0            0.0
low_adl_no_var                      0            0.0
ADL_first_score                     0            0.0
ADL_last_score                      0            0.0
ADL_diff_seq                        0            0.0
diff_has_denture                    0            0.0
last_has_denture                    0            0.0
意識總分Max                             0            0.0
使用呼吸輔具                              0            0.0
first_has_denture                   0            0.0
had_fall                            0            0.0
diff_has_feeding_tube               0            0.0
last_has_feeding_tube               0            0.0
first_has_feeding_tube              0            0.0
ADL_std                             0            0.0

In [ ]:
from sklearn.impute import SimpleImputer
from sklearn.base import clone
from sklearn.exceptions import NotFittedError
from sklearn.utils.validation import check_is_fitted
import plotly.subplots as sp

def evaluate_all_models_visual(models: dict, X_val, y_val):
    mean_fpr = np.linspace(0, 1, 100)
    fig_roc = go.Figure()
    results = []

    # 建立混淆矩陣子圖
    num_models = len(models)
    cols = 3
    rows = int(np.ceil(num_models / cols))

    fig_cm = sp.make_subplots(
        rows=rows, cols=cols,
        subplot_titles=list(models.keys()),
        horizontal_spacing=0.15,
        vertical_spacing=0.15
    )

    for i, (model_name, model) in enumerate(models.items()):
        print(f"🔍 Evaluating {model_name}...")

        # 嘗試使用原始資料
        X_input = X_val.copy()
        y_input = y_val

        # 若模型不支援 NaN，則補值
        try:
            # 嘗試呼叫 predict_proba
            _ = model.predict_proba(X_input)
        except ValueError as e:
            if "Input X contains NaN" in str(e):
                print(f"⚠️  {model_name} 不支援 NaN，自動補值中...")
                imputer = SimpleImputer(strategy="median")
                X_input = pd.DataFrame(imputer.fit_transform(X_input), columns=X_val.columns)
            else:
                raise e

        # 預測
        y_pred = model.predict(X_input)
        y_prob = model.predict_proba(X_input)[:, 1]

        # 指標
        acc = accuracy_score(y_input, y_pred)
        prec = precision_score(y_input, y_pred)
        rec = recall_score(y_input, y_pred)
        f1 = f1_score(y_input, y_pred)
        auc_val = roc_auc_score(y_input, y_prob)

        # ROC
        fpr, tpr, _ = roc_curve(y_input, y_prob)
        tpr_interp = np.interp(mean_fpr, fpr, tpr)
        tpr_interp[0], tpr_interp[-1] = 0.0, 1.0

        fig_roc.add_trace(go.Scatter(
            x=mean_fpr, y=tpr_interp,
            mode='lines',
            name=f"{model_name} (AUC={auc_val:.3f})"
        ))

        # 混淆矩陣
        cm = confusion_matrix(y_input, y_pred)
        row, col = i // cols + 1, i % cols + 1
        fig_cm.add_trace(
            go.Heatmap(
                z=cm,
                x=["Predicted Negative", "Predicted Positive"],
                y=["Actual Negative", "Actual Positive"],
                colorscale='Blues',
                text=cm,
                texttemplate="%{text}",
                showscale=False
            ),
            row=row, col=col
        )

        results.append({
            "Model": model_name,
            "Accuracy": acc,
            "Precision": prec,
            "Recall": rec,
            "F1 Score": f1,
            "ROC AUC": auc_val
        })

    # 隨機基準線
    fig_roc.add_trace(go.Scatter(
        x=[0, 1], y=[0, 1],
        mode='lines',
        line=dict(dash='dash'),
        name='Random Baseline'
    ))

    fig_roc.update_layout(
        title="ROC Curve Comparison",
        xaxis_title="False Positive Rate",
        yaxis_title="True Positive Rate",
        width=800,
        height=600
    )
    fig_roc.show()

    fig_cm.update_layout(
        title="Confusion Matrices of All Models",
        width=400 * cols,
        height=300 * rows,
        showlegend=False
    )
    fig_cm.show()

    # 指標表格
    df_result = pd.DataFrame(results)

    return df_result

In [ ]:
# 假設已經訓練完模型並存在 trained_models 中
evaluate_all_models_visual(trained_models, ex_X, ex_y)

🔍 Evaluating HybridXGBRF (Our Approach)...
🔍 Evaluating LogisticRegression (max_iter=200)...
🔍 Evaluating XGBClassifier...
🔍 Evaluating RandomForestClassifier...
🔍 Evaluating LogisticRegression (max_iter=1000)...
🔍 Evaluating Ridge...
🔍 Evaluating Lasso...
🔍 Evaluating Elastic...


Model  Accuracy  Precision    Recall  \
0          HybridXGBRF (Our Approach)  0.788137   0.845426  0.300955   
1   LogisticRegression (max_iter=200)  0.790503   0.824034  0.323414   
2                       XGBClassifier  0.788295   0.838213  0.305446   
3              RandomForestClassifier  0.778356   0.653595  0.449186   
4  LogisticRegression (max_iter=1000)  0.791765   0.825106  0.328467   
5                               Ridge  0.791765   0.829757  0.325660   
6                               Lasso  0.791608   0.826705  0.326783   
7                             Elastic  0.791765   0.827881  0.326783   

   F1 Score   ROC AUC  
0  0.443892  0.827786  
1  0.464516  0.789285  
2  0.447737  0.828064  
3  0.532446  0.792245  
4  0.469880  0.792450  
5  0.467742  0.791078  
6  0.468410  0.792160  
7  0.468599  0.791593

In [ ]:
# ===== 次族群實驗：年齡、ADL 變化、性別 =====
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

def _first_existing_column(df, candidates):
    for c in candidates:
        if c in df.columns:
            return c
    return None

def _make_sex_masks(df):
    """
    回傳 {'男性': mask, '女性': mask}；若無法判斷，回傳空 dict。
    支援：
      1) '性別' 欄位（值可能為 '男'/'女' 或 1/0/2…）
      2) one-hot 欄位 '性別_男' / '性別_女'
    """
    masks = {}
    if '性別_is_male' in df.columns:
        col = df['性別_is_male']
        # 嘗試各種常見標記
        male_mask = col.astype(str).str.contains('男') | (col == 1) | (col.astype(str).str.lower().isin(['m','male']))
        female_mask = col.astype(str).str.contains('女') | (col == 0) | (col.astype(str).str.lower().isin(['f','female']))
        if male_mask.any(): masks['男性'] = male_mask
        if female_mask.any(): masks['女性'] = female_mask
    else:
        male_col = _first_existing_column(df, ['性別_男','男','male','Male','M'])
        female_col = _first_existing_column(df, ['性別_女','女','female','Female','F'])
        if male_col is not None:
            masks['男性'] = df[male_col] == 1
        if female_col is not None:
            masks['女性'] = df[female_col] == 1
    return masks

def _make_adl_change_masks(df):
    """
    建立 ADL 變好/變差遮罩：
    'ADL_明顯惡化'（=0 視為變好，=1 變差）
    """
    masks = {}
    masks['ADL 變好'] = df['ADL_明顯惡化'] == 0
    masks['ADL 變差'] = df['ADL_明顯惡化'] == 1
    return masks

def _compute_metrics(y_true, y_prob, y_pred):
    return {
        'Accuracy': accuracy_score(y_true, y_pred),
        'Precision': precision_score(y_true, y_pred, zero_division=0),
        'Recall': recall_score(y_true, y_pred, zero_division=0),
        'F1': f1_score(y_true, y_pred, zero_division=0),
        'ROC AUC': roc_auc_score(y_true, y_prob) if len(np.unique(y_true)) > 1 else np.nan,
        'Support (n)': int(len(y_true)),
        'Positives (n)': int(y_true.sum())
    }

def evaluate_subgroups(models: dict, X_all: pd.DataFrame, y_all: pd.Series, raw_df_for_masks: pd.DataFrame):
    """
    models: 已訓練好的模型字典 trained_models
    X_all, y_all: 用於評估的特徵與標記（例如 ex_X, ex_y）
    raw_df_for_masks: 與 X_all 對齊、包含「年齡/ADL/性別」原始欄位的 DataFrame（例如 external）
    """
    # 年齡遮罩（需 '預估年齡'）
    subgroup_masks = {}
    if '預估年齡' in raw_df_for_masks.columns:
        subgroup_masks['年齡 > 85'] = raw_df_for_masks['預估年齡'] > 85
        subgroup_masks['年齡 <= 85'] = raw_df_for_masks['預估年齡'] <= 85
    else:
        print("⚠️ 找不到欄位『預估年齡』，跳過年齡分組。")

    # ADL 變化遮罩
    adl_masks = _make_adl_change_masks(raw_df_for_masks)
    if adl_masks:
        subgroup_masks.update(adl_masks)
    else:
        print("⚠️ 找不到可推算 ADL 變化的欄位，跳過 ADL 分組。")

    # 性別遮罩
    sex_masks = _make_sex_masks(raw_df_for_masks)
    if sex_masks:
        subgroup_masks.update(sex_masks)
    else:
        print("⚠️ 找不到可用的性別欄位，跳過性別分組。")

    rows = []
    for subgroup_name, mask in subgroup_masks.items():
        mask = mask.fillna(False).astype(bool)  # 安全轉型
        if mask.sum() == 0:
            print(f"⚠️ 次族群「{subgroup_name}」資料筆數為 0，略過。")
            continue

        X_sub = X_all.loc[mask]
        y_sub = y_all.loc[mask]

        # 若模型不支援 NaN，與你上面一致，統一補值策略（中位數）
        from sklearn.impute import SimpleImputer
        imputer = SimpleImputer(strategy="median")
        X_sub_imp = pd.DataFrame(imputer.fit_transform(X_sub), columns=X_sub.columns, index=X_sub.index)

        for model_name, model in models.items():
            # 預測
            y_prob = model.predict_proba(X_sub_imp)[:, 1]
            y_pred = (y_prob >= 0.5).astype(int)

            metrics = _compute_metrics(y_sub, y_prob, y_pred)
            metrics.update({'Subgroup': subgroup_name, 'Model': model_name})
            rows.append(metrics)

    result_df = pd.DataFrame(rows)
    if not result_df.empty:
        # 排序：先按次族群，再按 ROC AUC 由高到低
        result_df = result_df.sort_values(by=['Subgroup','ROC AUC'], ascending=[True, False]).reset_index(drop=True)
    return result_df

# === 執行：以外部驗證集為例 ===
subgroup_results = evaluate_subgroups(trained_models, ex_X, ex_y, external)
display(subgroup_results)

Accuracy  Precision    Recall        F1   ROC AUC  Support (n)  \
0   0.779505   0.838213  0.327908  0.471404  0.839347         5533   
1   0.779324   0.845426  0.323086  0.467510  0.838111         5533   
2   0.766853   0.650367  0.481013  0.553015  0.794909         5533   
3   0.780770   0.832836  0.336347  0.479176  0.790434         5533   
4   0.780589   0.832586  0.335744  0.478522  0.790392         5533   
5   0.780770   0.833832  0.335744  0.478728  0.789873         5533   
6   0.780770   0.833832  0.335744  0.478728  0.789427         5533   
7   0.780047   0.832831  0.333333  0.476108  0.786210         5533   
8   0.867246   0.692308  0.221311  0.335404  0.711018          806   
9   0.867246   0.714286  0.204918  0.318471  0.708861          806   
10  0.867246   0.714286  0.204918  0.318471  0.707063          806   
11  0.862283   0.657143  0.188525  0.292994  0.706584          806   
12  0.867246   0.741935  0.188525  0.300654  0.705457          806   
13  0.848635   0.000000  0.000000  0.000000  0.685055          806   
14  0.851117   0.600000  0.049180  0.090909  0.682449          806   
15  0.848635   0.000000  0.000000  0.000000  0.651520          806   
16  0.824454   0.805861  0.308123  0.445795  0.821830         3116   
17  0.824454   0.797153  0.313725  0.450251  0.818523         3116   
18  0.801990   0.604752  0.392157  0.475786  0.778588         3116   
19  0.824134   0.782313  0.322129  0.456349  0.776421         3116   
20  0.824454   0.781145  0.324930  0.458952  0.774975         3116   
21  0.824775   0.783784  0.324930  0.459406  0.774761         3116   
22  0.825096   0.786441  0.324930  0.459861  0.774138         3116   
23  0.825096   0.788396  0.323529  0.458788  0.773507         3116   
24  0.791543   0.828851  0.287045  0.426415  0.825394         4375   
25  0.791543   0.837093  0.282811  0.422785  0.823619         4375   
26  0.776457   0.621266  0.440305  0.515362  0.791381         4375   
27  0.794286   0.814318  0.308213  0.447174  0.791175         4375   
28  0.794514   0.817568  0.307367  0.446769  0.790969         4375   
29  0.794514   0.817568  0.307367  0.446769  0.790284         4375   
30  0.794514   0.819005  0.306520  0.446087  0.789641         4375   
31  0.793600   0.813063  0.305673  0.444308  0.786665         4375   
32  0.780550   0.859574  0.336667  0.483832  0.835904         1964   
33  0.781059   0.854167  0.341667  0.488095  0.833238         1964   
34  0.783605   0.843137  0.358333  0.502924  0.796681         1964   
35  0.786151   0.843511  0.368333  0.512761  0.794142         1964   
36  0.785132   0.842308  0.365000  0.509302  0.793583         1964   
37  0.785642   0.845560  0.365000  0.509895  0.793223         1964   
38  0.780041   0.710000  0.473333  0.568000  0.793120         1964   
39  0.785642   0.848249  0.363333  0.508751  0.792989         1964   
40  0.753335   0.869565  0.299906  0.445993  0.828236         3223   
41  0.753025   0.875346  0.296157  0.442577  0.827325         3223   
42  0.753956   0.677003  0.491097  0.569256  0.794586         3223   
43  0.760161   0.856796  0.330834  0.477350  0.791458         3223   
44  0.759541   0.857843  0.328022  0.474576  0.791044         3223   
45  0.759541   0.857843  0.328022  0.474576  0.790517         3223   
46  0.759541   0.859606  0.327085  0.473863  0.790067         3223   
47  0.757989   0.854321  0.324274  0.470109  0.781462         3223   

    Positives (n)  Subgroup                               Model  
0            1659    ADL 變好                       XGBClassifier  
1            1659    ADL 變好          HybridXGBRF (Our Approach)  
2            1659    ADL 變好              RandomForestClassifier  
3            1659    ADL 變好  LogisticRegression (max_iter=1000)  
4            1659    ADL 變好                               Lasso  
5            1659    ADL 變好                             Elastic  
6            1659    ADL 變好                               Ridge  
7            1659    ADL 變好   LogisticRegress

In [ ]:
# 若只想看主力模型
display(subgroup_results[subgroup_results['Model'] == 'HybridXGBRF (Our Approach)'])

Accuracy  Precision    Recall        F1   ROC AUC  Support (n)  \
1   0.779324   0.845426  0.323086  0.467510  0.838111         5533   
13  0.848635   0.000000  0.000000  0.000000  0.685055          806   
16  0.824454   0.805861  0.308123  0.445795  0.821830         3116   
25  0.791543   0.837093  0.282811  0.422785  0.823619         4375   
32  0.780550   0.859574  0.336667  0.483832  0.835904         1964   
41  0.753025   0.875346  0.296157  0.442577  0.827325         3223   

    Positives (n)  Subgroup                       Model  
1            1659    ADL 變好  HybridXGBRF (Our Approach)  
13            122    ADL 變差  HybridXGBRF (Our Approach)  
16            714        女性  HybridXGBRF (Our Approach)  
25           1181  年齡 <= 85  HybridXGBRF (Our Approach)  
32            600   年齡 > 85  HybridXGBRF (Our Approach)  
41           1067        男性  HybridXGBRF (Our Approach)

在 ADL 變差 這個子群組裡，雖然真實有 122 個 positive，但模型可能 完全沒預測出任何一個 positive（全部都預測為 negative）。

Precision = 0：因為 precision = TP / (TP + FP)，如果模型完全沒有預測 positive（TP = 0 且 FP = 0），precision 就定義為 0。

Recall = 0：因為 recall = TP / (TP + FN)，TP = 0，所以 recall 也變成 0。

F1 = 0：因為 F1 是 precision 與 recall 的調和平均，只要 precision 和 recall 其中之一是 0，F1 就會是 0。

In [ ]:
# ===== 複合次族群：年齡 × 性別 × ADL是否明顯惡化 =====
from itertools import product
import pandas as pd
from sklearn.impute import SimpleImputer

def evaluate_age_sex_composites(
    models: dict,
    X_all: pd.DataFrame,
    y_all: pd.Series,
    raw_df_for_masks: pd.DataFrame,
    age_threshold: int = 85,
    min_support: int = 10   # 次族群最少樣本數，太小就略過避免不穩定
):
    """
    針對「年齡(>threshold / <=threshold) × 性別(男性/女性) × ADL(變好/變差)」的交叉次族群做評估。
    會輸出每個模型在各交叉次族群的 Accuracy / Precision / Recall / F1 / ROC-AUC 等。
    依賴你已定義的: _make_sex_masks, _make_adl_change_masks, _compute_metrics。
    """
    if '預估年齡' not in raw_df_for_masks.columns:
        print("⚠️ 找不到欄位『預估年齡』，無法建立年齡遮罩。")
        return pd.DataFrame()

    # 年齡 bins
    age_masks = {
        f'年齡 > {age_threshold}': (raw_df_for_masks['預估年齡'] > age_threshold),
        f'年齡 <= {age_threshold}': (raw_df_for_masks['預估年齡'] <= age_threshold),
    }

    # 性別 masks（沿用你上面的 _make_sex_masks）
    sex_masks = _make_sex_masks(raw_df_for_masks)
    if not sex_masks:
        print("⚠️ 找不到可用的性別欄位，無法建立性別遮罩。")
        return pd.DataFrame()

    # ADL 變化 masks（沿用你上面的 _make_adl_change_masks）
    if 'ADL_明顯惡化' not in raw_df_for_masks.columns:
        print("⚠️ 找不到欄位『ADL_明顯惡化』，無法建立 ADL 遮罩。")
        return pd.DataFrame()
    adl_masks = _make_adl_change_masks(raw_df_for_masks)  # {'ADL 變好': mask, 'ADL 變差': mask}

    rows = []
    for (age_name, age_mask), (sex_name, sex_mask), (adl_name, adl_mask) in product(
        age_masks.items(), sex_masks.items(), adl_masks.items()
    ):
        combo_name = f"{age_name} & {sex_name} & {adl_name}"
        mask = (
            age_mask.fillna(False).astype(bool)
            & sex_mask.fillna(False).astype(bool)
            & adl_mask.fillna(False).astype(bool)
        )
        n = int(mask.sum())
        if n < min_support:
            print(f"ℹ️ 複合次族群「{combo_name}」樣本數 {n} < min_support={min_support}，略過。")
            continue

        X_sub = X_all.loc[mask]
        y_sub = y_all.loc[mask]

        # 與你現有策略一致：補缺失值（中位數）
        imputer = SimpleImputer(strategy="median")
        X_sub_imp = pd.DataFrame(imputer.fit_transform(X_sub), columns=X_sub.columns, index=X_sub.index)

        for model_name, model in models.items():
            y_prob = model.predict_proba(X_sub_imp)[:, 1]
            y_pred = (y_prob >= 0.5).astype(int)

            metrics = _compute_metrics(y_sub, y_prob, y_pred)
            metrics.update({'Subgroup': combo_name, 'Model': model_name})
            rows.append(metrics)

    result_df = pd.DataFrame(rows)
    if not result_df.empty:
        result_df['Prevalence'] = result_df['Positives (n)'] / result_df['Support (n)']
        result_df = result_df.sort_values(by=['Subgroup', 'ROC AUC'], ascending=[True, False]).reset_index(drop=True)
    return result_df

In [ ]:
age_sex_results = evaluate_age_sex_composites(
    trained_models, ex_X, ex_y, external,
    age_threshold=85,
    min_support=10
)
display(age_sex_results)

pivot_auc = age_sex_results.pivot_table(index='Subgroup', columns='Model', values='ROC AUC')
display(pivot_auc)

Accuracy  Precision    Recall        F1   ROC AUC  Support (n)  \
0   0.832725   0.805755  0.311111  0.448898  0.837258         1644   
1   0.832117   0.791667  0.316667  0.452381  0.836664         1644   
2   0.799270   0.556391  0.411111  0.472843  0.787451         1644   
3   0.831509   0.786207  0.316667  0.451485  0.776698         1644   
4   0.831509   0.786207  0.316667  0.451485  0.773157         1644   
..       ...        ...       ...       ...       ...          ...   
59  0.864865   0.875000  0.333333  0.482759  0.695767          111   
60  0.873874   0.888889  0.380952  0.533333  0.695767          111   
61  0.864865   0.875000  0.333333  0.482759  0.694180          111   
62  0.810811   0.000000  0.000000  0.000000  0.658201          111   
63  0.810811   0.000000  0.000000  0.000000  0.628042          111   

    Positives (n)                Subgroup                              Model  \
0             360  年齡 <= 85 & 女性 & ADL 變好         HybridXGBRF (Our Approach)   
1             360  年齡 <= 85 & 女性 & ADL 變好                      XGBClassifier   
2             360  年齡 <= 85 & 女性 & ADL 變好             RandomForestClassifier   
3             360  年齡 <= 85 & 女性 & ADL 變好  LogisticRegression (max_iter=200)   
4             360  年齡 <= 85 & 女性 & ADL 變好                              Lasso   
..            ...                     ...                                ...   
59             21   年齡 > 85 & 男性 & ADL 變差  LogisticRegression (max_iter=200)   
60             21   年齡 > 85 & 男性 & ADL 變差                            Elastic   
61             21   年齡 > 85 & 男性 & ADL 變差                              Ridge   
62             21   年齡 > 85 & 男性 & ADL 變差         HybridXGBRF (Our Approach)   
63             21   年齡 > 85 & 男性 & ADL 變差                      XGBClassifier   

    Prevalence  
0     0.218978  
1     0.218978  
2     0.218978  
3     0.218978  
4     0.218978  
..         ...  
59    0.189189  
60    0.189189  
61    0.189189  
62    0.189189  
63    0.189189  

[64 rows x 10 columns]

Model                    Elastic  HybridXGBRF (Our Approach)     Lasso  \
Subgroup                                                                 
年齡 <= 85 & 女性 & ADL 變好  0.772702                    0.837258  0.773157   
年齡 <= 85 & 女性 & ADL 變差  0.675036                    0.680880  0.676479   
年齡 <= 85 & 男性 & ADL 變好  0.779385                    0.828989  0.780037   
年齡 <= 85 & 男性 & ADL 變差  0.741495                    0.694361  0.743925   
年齡 > 85 & 女性 & ADL 變好   0.761259                    0.823833  0.761667   
年齡 > 85 & 女性 & ADL 變差   0.653725                    0.676275  0.656863   
年齡 > 85 & 男性 & ADL 變好   0.799720                    0.853423  0.800208   
年齡 > 85 & 男性 & ADL 變差   0.695767                    0.658201  0.696825   

Model                   LogisticRegression (max_iter=1000)  \
Subgroup                                                     
年齡 <= 85 & 女性 & ADL 變好                            0.773016   
年齡 <= 85 & 女性 & ADL 變差                            0.679365   
年齡 <= 85 & 男性 & ADL 變好                            0.780257   
年齡 <= 85 & 男性 & ADL 變差                            0.746690   
年齡 > 85 & 女性 & ADL 變好                             0.761745   
年齡 > 85 & 女性 & ADL 變差                             0.658431   
年齡 > 85 & 男性 & ADL 變好                             0.801976   
年齡 > 85 & 男性 & ADL 變差                             0.704233   

Model                   LogisticRegression (max_iter=200)  \
Subgroup                                                    
年齡 <= 85 & 女性 & ADL 變好                           0.776698   
年齡 <= 85 & 女性 & ADL 變差                           0.675613   
年齡 <= 85 & 男性 & ADL 變好                           0.767111   
年齡 <= 85 & 男性 & ADL 變差                           0.733618   
年齡 > 85 & 女性 & ADL 變好                            0.766736   
年齡 > 85 & 女性 & ADL 變差                            0.656863   
年齡 > 85 & 男性 & ADL 變好                            0.799106   
年齡 > 85 & 男性 & ADL 變差                            0.695767   

Model                   RandomForestClassifier     Ridge  XGBClassifier  
Subgroup                                                                 
年齡 <= 85 & 女性 & ADL 變好                0.787451  0.772002       0.836664  
年齡 <= 85 & 女性 & ADL 變差                0.724531  0.673449       0.649639  
年齡 <= 85 & 男性 & ADL 變好                0.790716  0.778832       0.832650  
年齡 <= 85 & 男性 & ADL 變差                0.655941  0.739735       0.650075  
年齡 > 85 & 女性 & ADL 變好                 0.767072  0.760812       0.821316  
年齡 > 85 & 女性 & ADL 變差                 0.635098  0.650196       0.616471  
年齡 > 85 & 男性 & ADL 變好                 0.815198  0.799646       0.850394  
年齡 > 85 & 男性 & ADL 變差                 0.706085  0.694180       0.628042

In [ ]:
# 每個模型在哪個次族群表現最好（同樣以 ROC AUC 為主）
def best_subgroup_per_model(results_df: pd.DataFrame,
                            primary='ROC AUC',
                            tie_breakers=('F1','Recall','Precision','Accuracy','Support (n)')):
    sort_cols = ['Model', primary, *tie_breakers]
    sort_asc  = [True, False, *([False]*len(tie_breakers))]
    df_sorted = results_df.sort_values(by=sort_cols, ascending=sort_asc)
    best_df = df_sorted.groupby('Model', as_index=False).head(1).reset_index(drop=True)
    return best_df

best_subgroup_each_model = best_subgroup_per_model(age_sex_results)
display(best_subgroup_each_model)

Accuracy  Precision    Recall        F1   ROC AUC  Support (n)  \
0  0.730769   0.916667  0.383721  0.540984  0.799720          624   
1  0.725962   0.930693  0.364341  0.523677  0.853423          624   
2  0.730769   0.916667  0.383721  0.540984  0.800208          624   
3  0.730769   0.916667  0.383721  0.540984  0.801976          624   
4  0.727564   0.915094  0.375969  0.532967  0.799106          624   
5  0.751603   0.748792  0.600775  0.666667  0.815198          624   
6  0.730769   0.916667  0.383721  0.540984  0.799646          624   
7  0.725962   0.922330  0.368217  0.526316  0.850394          624   

   Positives (n)               Subgroup                               Model  \
0            258  年齡 > 85 & 男性 & ADL 變好                             Elastic   
1            258  年齡 > 85 & 男性 & ADL 變好          HybridXGBRF (Our Approach)   
2            258  年齡 > 85 & 男性 & ADL 變好                               Lasso   
3            258  年齡 > 85 & 男性 & ADL 變好  LogisticRegression (max_iter=1000)   
4            258  年齡 > 85 & 男性 & ADL 變好   LogisticRegression (max_iter=200)   
5            258  年齡 > 85 & 男性 & ADL 變好              RandomForestClassifier   
6            258  年齡 > 85 & 男性 & ADL 變好                               Ridge   
7            258  年齡 > 85 & 男性 & ADL 變好                       XGBClassifier   

   Prevalence  
0    0.413462  
1    0.413462  
2    0.413462  
3    0.413462  
4    0.413462  
5    0.413462  
6    0.413462  
7    0.413462

In [ ]:
# 直接列出「整體表現最高的 (Subgroup, Model) Top-K」
def top_k_overall(results_df: pd.DataFrame, k=10,
                  primary='ROC AUC',
                  tie_breakers=('F1','Recall','Precision','Accuracy','Support (n)')):
    sort_cols = [primary, *tie_breakers]
    sort_asc  = [False, *([False]*len(tie_breakers))]
    return results_df.sort_values(by=sort_cols, ascending=sort_asc).head(k).reset_index(drop=True)

top10 = top_k_overall(age_sex_results, k=10)
display(top10)

Accuracy  Precision    Recall        F1   ROC AUC  Support (n)  \
0  0.725962   0.930693  0.364341  0.523677  0.853423          624   
1  0.725962   0.922330  0.368217  0.526316  0.850394          624   
2  0.832725   0.805755  0.311111  0.448898  0.837258         1644   
3  0.832117   0.791667  0.316667  0.452381  0.836664         1644   
4  0.749433   0.849057  0.305292  0.449102  0.832650         2203   
5  0.748979   0.853846  0.301221  0.445336  0.828989         2203   
6  0.790960   0.805970  0.355263  0.493151  0.823833         1062   
7  0.791902   0.802920  0.361842  0.498866  0.821316         1062   
8  0.751603   0.748792  0.600775  0.666667  0.815198          624   
9  0.730769   0.916667  0.383721  0.540984  0.801976          624   

   Positives (n)                Subgroup                               Model  \
0            258   年齡 > 85 & 男性 & ADL 變好          HybridXGBRF (Our Approach)   
1            258   年齡 > 85 & 男性 & ADL 變好                       XGBClassifier   
2            360  年齡 <= 85 & 女性 & ADL 變好          HybridXGBRF (Our Approach)   
3            360  年齡 <= 85 & 女性 & ADL 變好                       XGBClassifier   
4            737  年齡 <= 85 & 男性 & ADL 變好                       XGBClassifier   
5            737  年齡 <= 85 & 男性 & ADL 變好          HybridXGBRF (Our Approach)   
6            304   年齡 > 85 & 女性 & ADL 變好          HybridXGBRF (Our Approach)   
7            304   年齡 > 85 & 女性 & ADL 變好                       XGBClassifier   
8            258   年齡 > 85 & 男性 & ADL 變好              RandomForestClassifier   
9            258   年齡 > 85 & 男性 & ADL 變好  LogisticRegression (max_iter=1000)   

   Prevalence  
0    0.413462  
1    0.413462  
2    0.218978  
3    0.218978  
4    0.334544  
5    0.334544  
6    0.286252  
7    0.286252  
8    0.413462  
9    0.413462

In [ ]:
# 每個次族群下，表現最好的模型（以 ROC AUC 為主，F1/Recall/Precision/Accuracy/Support 作為平手時的次序）
def best_model_per_subgroup(results_df: pd.DataFrame,
                            primary='ROC AUC',
                            tie_breakers=('F1','Recall','Precision','Accuracy','Support (n)')):
    sort_cols = ['Subgroup', primary, *tie_breakers]
    sort_asc  = [True, False, *([False]*len(tie_breakers))]
    df_sorted = results_df.sort_values(by=sort_cols, ascending=sort_asc)
    # 取每個 Subgroup 的第一列（即最佳模型）
    best_df = df_sorted.groupby('Subgroup', as_index=False).head(1).reset_index(drop=True)
    return best_df

best_by_subgroup = best_model_per_subgroup(age_sex_results)
display(best_by_subgroup)

Accuracy  Precision    Recall        F1   ROC AUC  Support (n)  \
0  0.832725   0.805755  0.311111  0.448898  0.837258         1644   
1  0.872428   0.750000  0.090909  0.162162  0.724531          243   
2  0.749433   0.849057  0.305292  0.449102  0.832650         2203   
3  0.842105   0.687500  0.215686  0.328358  0.746690          285   
4  0.790960   0.805970  0.355263  0.493151  0.823833         1062   
5  0.898204   0.000000  0.000000  0.000000  0.676275          167   
6  0.725962   0.930693  0.364341  0.523677  0.853423          624   
7  0.810811   0.000000  0.000000  0.000000  0.706085          111   

   Positives (n)                Subgroup                               Model  \
0            360  年齡 <= 85 & 女性 & ADL 變好          HybridXGBRF (Our Approach)   
1             33  年齡 <= 85 & 女性 & ADL 變差              RandomForestClassifier   
2            737  年齡 <= 85 & 男性 & ADL 變好                       XGBClassifier   
3             51  年齡 <= 85 & 男性 & ADL 變差  LogisticRegression (max_iter=1000)   
4            304   年齡 > 85 & 女性 & ADL 變好          HybridXGBRF (Our Approach)   
5             17   年齡 > 85 & 女性 & ADL 變差          HybridXGBRF (Our Approach)   
6            258   年齡 > 85 & 男性 & ADL 變好          HybridXGBRF (Our Approach)   
7             21   年齡 > 85 & 男性 & ADL 變差              RandomForestClassifier   

   Prevalence  
0    0.218978  
1    0.135802  
2    0.334544  
3    0.178947  
4    0.286252  
5    0.101796  
6    0.413462  
7    0.189189